### PR TITLE
Refactor Lua manager to use std::vector for m_VirtualMachines

### PR DESCRIPTION
**Pull request by Pirulax**
Improves performance by using `std::vector` instead of a list.
as you can see from the commits, I first opted to use `CFastHashSet`, but, it's unnecessary, since we only use `find()` once. (When deleting the VM).
Most of the time we iterate thru the vector (`DoPulse()`), where the vector clearly wins in any scenario, and by a big margin.

Tldr: use `vector` instead of `list`, to make things faster, for the expense of `RemoveVM` getting slower.

### CODE REVIEW

**Comment by Pirulax** [comment]
I tested the server, works fine. The client should work just as fine.

File: Client/mods/deathmatch/logic/lua/CLuaManager.cpp [hunk 1]
```diff
@@ -72,7 +71,9 @@ bool CLuaManager::RemoveVirtualMachine(CLuaMain* pLuaMain)
         m_pRegisteredCommands->CleanUpForVM(pLuaMain);
 
         // Remove it from our list
-        m_virtualMachines.remove(pLuaMain);
+        const auto iter = std::find(m_VirtualMachines.begin(), m_VirtualMachines.end(), pLuaMain);
+        assert(iter != m_VirtualMachines.end()); // Make sure it exists
```
Comment (qaisjp) [inline Client/mods/deathmatch/logic/lua/CLuaManager.cpp]: Our auto-formatting does unusual things to inline comments, and this comment position is too sane, so I think the code is formatted incorrectly. Try formatting selection.

File: Client/mods/deathmatch/logic/lua/CLuaManager.cpp [hunk 2]
```diff
@@ -130,23 +128,25 @@ CLuaMain* CLuaManager::GetVirtualMachine(lua_State* luaVM)
...
 
     // Find a matching VM in our map
-    CLuaMain* pLuaMain = MapFindRef(m_VirtualMachineMap, luaVM);
-    if (pLuaMain)
-        return pLuaMain;
+    {
+        CLuaMain* pLuaMain = MapFindRef(m_VirtualMachineMap, luaVM);
+        if (pLuaMain)
+            return pLuaMain;
+    }
```
Comment (qaisjp) [inline Client/mods/deathmatch/logic/lua/CLuaManager.cpp]:
```suggestion
    CLuaMain* pLuaMain = MapFindRef(m_VirtualMachineMap, luaVM);
    if (pLuaMain)
        return pLuaMain;
```
Is the change in indentation+scoping necessary here?

File: Client/mods/deathmatch/logic/lua/CLuaManager.h [hunk 3]
```diff
@@ -33,8 +33,7 @@ class CLuaManager
     void      OnLuaMainOpenVM(CLuaMain* pLuaMain, lua_State* luaVM);
     void      OnLuaMainCloseVM(CLuaMain* pLuaMain, lua_State* luaVM);
 
-    std::list<CLuaMain*>::const_iterator IterBegin() { return m_virtualMachines.begin(); };
-    std::list<CLuaMain*>::const_iterator IterEnd() { return m_virtualMachines.end(); };
+    auto GetLuaMains() const { return m_VirtualMachines; }
```
Comment (qaisjp) [inline Client/mods/deathmatch/logic/lua/CLuaManager.h]: Why did you pick a return `auto` here? Seems like an unusual place to use it, especially since `m_VirtualMachines` doesn't have a particularly difficult type to write. I'm happy to keep `auto` if you think we should keep it, though.

File: Server/mods/deathmatch/logic/lua/CLuaManager.cpp [hunk 4]
```diff
@@ -79,12 +79,13 @@ bool CLuaManager::RemoveVirtualMachine(CLuaMain* pLuaMain)
...
         // Delete it unless it is already
         if (!pLuaMain->BeingDeleted())
-        {
             delete pLuaMain;
-        }
 
         // Remove it from our list
-        m_virtualMachines.remove(pLuaMain);
+        const auto iter = std::find(m_VirtualMachines.begin(), m_VirtualMachines.end(), pLuaMain);
+        assert(iter != m_VirtualMachines.end()); // Make sure it exists
```
Comment (qaisjp) [inline Server/mods/deathmatch/logic/lua/CLuaManager.cpp]: Format selection, as mentioned above

File: Server/mods/deathmatch/logic/lua/CLuaModule.cpp [hunk 5]
```diff
@@ -190,19 +190,17 @@ void CLuaModule::_RegisterFunctions(lua_State* luaVM)
...
+    for (CLuaMain* luaMain : m_pLuaModuleManager->GetLuaManager()->GetLuaMains())
     {
-        lua_State*                luaVM = (*liter)->GetVM();
-        vector<SString>::iterator iter = m_Functions.begin();
-        for (; iter != m_Functions.end(); ++iter)
+        lua_State* luaVM = luaMain->GetVM();
+        for (const SString& strFuncName : m_Functions)
         {
-            // points function to nill
+            // Remove function global variable
```
Comment (qaisjp) [inline Server/mods/deathmatch/logic/lua/CLuaModule.cpp]:
```suggestion
            // Set function global variable to nil
```

File: Client/mods/deathmatch/logic/lua/CLuaManager.h [hunk 6]
```diff
@@ -52,6 +51,6 @@ class CLuaManager
     CRegisteredCommands* m_pRegisteredCommands;
 
     CFastHashMap<lua_State*, CLuaMain*> m_VirtualMachineMap;
-    std::list<CLuaMain*>                m_virtualMachines;
+    std::vector<CLuaMain*>              m_VirtualMachines;
```
Comment (qaisjp) [inline Client/mods/deathmatch/logic/lua/CLuaManager.h:54]: You are changing type from `std::list` to `std::vector`.

http://www.cplusplus.com/reference/vector/vector/erase/ says this:

> Because vectors use an array as their underlying storage, erasing elements in positions other than the vector end causes the container to relocate all the elements after the segment erased to their new positions. This is generally an inefficient operation compared to the one performed for the same operation by other kinds of sequence containers (such as list or forward_list).

Have you taken this into consideration?

**Review by qaisjp — CHANGES_REQUESTED** [review]
Left a few comments, mostly just a few small nits and two questions. It looks fine, otherwise.

File: Client/mods/deathmatch/logic/lua/CLuaManager.h (hunk 3 above)
Comment (Pirulax) [reply Client/mods/deathmatch/logic/lua/CLuaManager.h]: Actually I shouldve returned `const auto&`.
The reason is that it doesnt matter what the type of `m_VirtualMachines` is. Its known to be an STL container, and I find this easier to read and maintain than:
```cpp
const std::vector<CLuaMain*>& GetLuaMains() const { return m_VirtualMachines; }
```

File: Client/mods/deathmatch/logic/lua/CLuaManager.h (hunk 6 above)
Comment (Pirulax) [reply Client/mods/deathmatch/logic/lua/CLuaManager.h:54]: I did.
But, `std list` has a huge overhead on small datatypes, and heap allocates every.
While deliting from a list is `O(n)`, in reality that `O(n)` is far slower than vectors `O(n^2)`, because while std vector will easily be cached, iterating thru it is very fast, compared to std::list, thanks to the CPU cache.
Also, copying it is really fast as well, since we're talking about pointers, which are 4 bytes. Even in the worst case, this will take maybe n * 10 clock cycles to copy, which is very fast, compared to `std::list` where we need to wait a lot more.

Tldr, yes I did, and std::vector will always be faster (if `sizeof(T)` is small (like a pointer is 4 bytes))

File: Client/mods/deathmatch/logic/lua/CLuaManager.h (hunk 3 above)
Comment (qaisjp) [reply Client/mods/deathmatch/logic/lua/CLuaManager.h]: Okay

File: Client/mods/deathmatch/logic/lua/CLuaManager.h (hunk 6 above)
Comment (qaisjp) [reply Client/mods/deathmatch/logic/lua/CLuaManager.h:54]: Okay 👍

File: Server/mods/deathmatch/logic/lua/CLuaManager.cpp [hunk 7]
```diff
@@ -93,30 +94,29 @@ bool CLuaManager::RemoveVirtualMachine(CLuaMain* pLuaMain)
 
 void CLuaManager::OnLuaMainOpenVM(CLuaMain* pLuaMain, lua_State* luaVM)
 {
-    MapSet(m_VirtualMachineMap, pLuaMain->GetVirtualMachine(), pLuaMain);
```
Comment (qaisjp) [inline Server/mods/deathmatch/logic/lua/CLuaManager.cpp:96]: It seems like MapSet and MapRemove is the standard way to interact with CFastHashMap. What's the purpose of the change here and in OnLuaMainCloseVM?

Comment (Pirulax) [reply Server/mods/deathmatch/logic/lua/CLuaManager.cpp:96]: Because `MapSet`, and `MapInsert` as useless. Their only purpose was to provide a better interface for setting/insertion, but since C++11 there are initializer lists, which make those functions totally useless.
That is, pre c++11:
```cpp
m_VirtualMachineMap.insert(std::pair<....>(...))
``` 
Now:
```cpp
m_VirtualMachineMap.insert({...})
```

As for `MapSet`.. I really dont see why it exists in the first place, it is just a simple:
```cpp
m_VirtualMachineMap[key] = value
```

I strongly disagree with the usage of these functions in new code.

Comment (qaisjp) [reply Server/mods/deathmatch/logic/lua/CLuaManager.cpp:96]: Isn't `m_VirtualMachineMap[key] = value` cleaner than `m_VirtualMachineMap.insert({key, value})`?

Comment (Pirulax) [reply Server/mods/deathmatch/logic/lua/CLuaManager.cpp:96]: It is, thanks!

**Comment by Pirulax** [comment]
TODO: (Theory) List was originally used because it might modify itself when deleting the VM or something.. Must check.
Well, had a quick look: It might happen if a timer callback stops a resource. Sadly this can't be solved easily.
